src/anal_ghidra.cpp

```diff
@@ -0,0 +1,3343 @@
+		}
+		if(anal_op->val && anal_op->val != -1)
+			std::cerr << " val: " << anal_op->val << std::endl;
+		else
+		{
+			if(anal_op->dst)
+			{
+				std::cerr << " dst: ";
+				char *tmp = r_anal_value_to_string(anal_op->dst);
+				std::cerr << tmp;
+				free(tmp);
+			}
+			if(anal_op->src[0])
+			{
+				std::cerr << " in0: ";
+				char *tmp = r_anal_value_to_string(anal_op->src[0]);
+				std::cerr << tmp;
+				free(tmp);
+			}
+			if(anal_op->src[1])
+			{
+				std::cerr << " in1: ";
+				char *tmp = r_anal_value_to_string(anal_op->src[1]);
+				std::cerr << tmp;
+				free(tmp);
+			}
+			std::cerr << std::endl;
+		}
+#endif
+	}
+
+	if(mask & R_ANAL_OP_MASK_ESIL)
+		sleigh_esil(a, anal_op, addr, data, len, pcode_slg.pcodes);
+
+	return anal_op->size;
+}
+
+/*
+ * By 2020-05-24, there are 17 kinds of group of registers in SLEIGH.
+ * I map them to r_reg.h's RRegisterType:
+ * R_REG_TYPE_XMM:
+ * R_REG_TYPE_SEG:
+ * R_REG_TYPE_DRX: DEBUG
+ * R_REG_TYPE_FPU: ST FPU
+ * R_REG_TYPE_MMX: MMX
+ * R_REG_TYPE_YMM: AVX VSX
+ * R_REG_TYPE_FLG: FLAGS Flags
+ * R_REG_TYPE_GPR: PC Cx DCR STATUS SVE CONTROL SPR SPR_UNNAMED Alt NEON
+ */
+static const char *r_reg_type_arr[] = {"PC",  "Cx",          "DCR", "STATUS", "SVE",   "CONTROL",
+                                "SPR", "SPR_UNNAMED", "Alt", "NEON",   "FLAGS", "Flags",
+                                "AVX", "MMX",         "ST",  "FPU",    "DEBUG", "VSX", nullptr};
+static const char *r_reg_string_arr[] = {"gpr", "gpr", "gpr", "gpr", "gpr", "gpr",
+                                  "gpr", "gpr", "gpr", "gpr", "flg", "flg",
+                                  "ymm", "mmx", "fpu", "fpu", "drx", "ymm", nullptr};
+
+static int get_reg_type(const std::string &name)
+{
+	auto p = sanal.reg_mapping.cbegin();
+	for(; p != sanal.reg_mapping.cend() && p->second != name; ++p) {}
+	if(p == sanal.reg_mapping.cend())
+		throw LowlevelError("get_reg_type: reg doesn't exist.");
+
+	const std::string &group = sanal.reg_group[p->first];
+
+	if(group.empty())
+		return R_REG_TYPE_GPR;
+
+	for(size_t i = 0; r_reg_type_arr[i]; i++)
+	{
+		if(group == r_reg_type_arr[i])
+		{
+			const char *curr = r_reg_string_arr[i];
+			switch(curr[0] | curr[1] << 8)
+			{
+				case 'g' | 'p' << 8: return R_REG_TYPE_GPR;
+				case 'd' | 'r' << 8: return R_REG_TYPE_DRX;
+				case 'f' | 'p' << 8: return R_REG_TYPE_FPU;
+				case 'm' | 'm' << 8: return R_REG_TYPE_MMX;
+				case 'x' | 'm' << 8: return R_REG_TYPE_XMM;
+				case 'y' | 'm' << 8: return R_REG_TYPE_YMM;
+				case 'f' | 'l' << 8: return R_REG_TYPE_FLG;
+				case 's' | 'e' << 8: return R_REG_TYPE_SEG;
+			}
+		}
+	}
+
+	return -1;
+}
+
+static void append_hardcoded_regs(std::stringstream &buf, const std::string &arch, bool little,
+                                  int bits)
+{
+	if(arch.size() < 3)
+		throw LowlevelError("append_hardcoded_regs: Unexpected arch name.");
+
+	switch(arch[0] | arch[1] << 8 | arch[2] << 16)
```

Comment:
I don't think you have to optimize the switch on the string this hard. As the nicest solution I would recommend to do something like here:
https://github.com/radareorg/r2ghidra-dec/blob/8e576eeadc211de4ac8d8c759cc368fa48cdfa99/src/ArchMap.cpp#L67-L79
Maybe you can even re-use the Mapper classes there if you move them to the header and replace `RCore *` here by a template parameter:
https://github.com/radareorg/r2ghidra-dec/blob/8e576eeadc211de4ac8d8c759cc368fa48cdfa99/src/ArchMap.cpp#L10-L14

But alternatively, a simple if/else would work too if you don't want to over-engineer it too much.

Reply:
Also, it should not fail completely if the arch is not known. You should still be able to use the plugin at least with unrecognized BP and SN.

Reply:
I will save this for next pr. Since current map is got from existing RAnal plugin source code, I think their result is not good enough. There must be something missing. I prepare to redo this in the future.

Reply:
I agree with both. Yes no need to optimize that and no need to make a new commit to address this now.